### PR TITLE
[stdlib] Speeding up typechecking for complex expressions with integers

### LINKED ISSUE #47286: [SR-4709] SRP source breakage: error: expression was too complex
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-4709      |
   |Radar             | rdar://problem/31845712         |
   |Original Reporter | @lplarson      |
   |Type              | Bug    |

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       | Compiler    |
   |Labels            | Bug, SourceCompatibility, TypeChecker        |
   |Assignee          | None      |
   |Priority          | Medium      |

   

   md5: 2e50e5ee9e8bcd730cb2398f9fb6cf42

  </details>

**is blocked by**:  
* [SR-4864](https://bugs.swift.org/browse/SR-4864) error: expression was too complex to be solved in reasonable time  

**Issue Description:**

To reproduce:  
1. Install latest GM/Beta Xcode  
2. $ git clone git@github.com:apple/swift-source-compat-suite.git  
3. $ cd swift-source-compat-suite

To build Swift from scratch before testing:  
4. $ ./reproduce.py master --project-path SRP

Or if you've already built Swift:  
4. $ ./reproduce.py master --project-path SRP --swiftc path/to/swiftc

SipHasher.swift:161:17: error: expression was too complex to be solved in reasonable time; consider breaking up the expression into distinct sub-expressions  
\| (UInt64(buffer\[i + 7\]) \<\< 56)

### PR DESCRIPTION
**Pull request by moiseev**
Utilizing the shortcut in the compiler to make expression with heavy use of bitwise operators and shifts compile much faster, especially in Swift 3 mode.

Fixes #47286, and also <rdar://problem/31543153>.

### CODE REVIEW

**Comment by moiseev** [comment]
@swift-ci Please smoke benchmark

**Comment by moiseev** [comment]
@swift-ci Please test

**Comment by moiseev** [comment]
@swift-ci Please Test Source Compatibility

File: benchmark/single-source/ByteSwap.swift [hunk 1]
```diff
@@ -18,15 +18,25 @@ import TestsUtils
...
-  var result = ((a & 0x00000000000000FF) << 56)
-  result |=    ((a & 0x000000000000FF00) << 40)
-  result |=    ((a & 0x0000000000FF0000) << 24)
-  result |=    ((a & 0x00000000FF000000) <<  8)
-  result |=    ((a & 0x000000FF00000000) >>  8)
-  result |=    ((a & 0x0000FF0000000000) >> 24)
-  result |=    ((a & 0x00FF000000000000) >> 40)
-  result |=    ((a & 0xFF00000000000000) >> 56)
-  return result
+#if swift(>=4)
```
Comment (jrose-apple) [inline benchmark/single-source/ByteSwap.swift]: Should this be `swift(>=4)` or `swift(>=3.2)`?

Comment (moiseev) [inline benchmark/single-source/ByteSwap.swift]: `>=4` covers the compatibility mode, so the former. Not that it matters much in the benchmark, that's always built in Swift 4 mode.

**Comment by moiseev** [comment]
@swift-ci Please test

**Comment by moiseev** [comment]
@swift-ci Please smoke benchmark

**Comment by moiseev** [comment]
@swift-ci Please Test Source Compatibility

File: benchmark/single-source/ByteSwap.swift (hunk 1 above)
Comment (jrose-apple) [inline benchmark/single-source/ByteSwap.swift]: Do you mean you want the compatibility mode to take the old path or the new? It will take the old.

Comment (moiseev) [inline benchmark/single-source/ByteSwap.swift]: Yes, that is what I want it to do. My phrasing leaves a lot to be desired. Compatiblity mode should take the same path as Swift 3 compiler would.

**Comment by dabrahams** [comment]
The need for these changes makes me 😢 but glad you figured this out

**Pull request merged**